File: 572-subtree-of-another-tree/572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        
        def dfs(root, sRoot):
                
            if not root and not sRoot:
                return True
            if not root or not sRoot:
                return False
            
            return root.val == sRoot.val and dfs(root.left, sRoot.left) and dfs(root.right, sRoot.right)
        
                    
        
        return dfs(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: 572-subtree-of-another-tree/572-subtree-of-another-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool dfs(struct TreeNode * root, struct TreeNode * subRoot){
    if(root == NULL && subRoot == NULL)
        return true;
    if(root == NULL || subRoot == NULL)
        return false;
    
    return (root->val == subRoot->val) && dfs(root->left, subRoot->left) && dfs(root->right, subRoot->right);
    
}

bool isSubtree(struct TreeNode* root, struct TreeNode* subRoot){
    if(root == NULL)
        return false;
    return dfs(root, subRoot) || isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
}